ibis/common/tests/test_deferred.py

```diff
@@ -602,3 +583,25 @@ def test_deferred_namespace(table):
 def test_custom_deferred_repr(table):
     expr = _.x + table.a
     assert repr(expr) == "(_.x + <column[int]>)"
+
+
+def test_deferred_hashable():
+    assert bool(_)
+    assert bool(_.a)
+    assert bool(_.a == _.a)
```

Comment:
Does this make the use of eq expressions as predicates confusing?

Reply:
I think it will be confusing to support the eq expression as well as the default bool support. In some cases the == will result in an expression, but if used in any API calling bool, it will be converted to a bool. Can we continue to raise?

Reply:
> Can we continue to raise?

Nope, otherwise `dct[_.a]` is going to raise. I mainly put up the PR so we can discuss this in details :)

Reply:
In that case I'm -0.5 on making deferred objects hashable.

Why should we take on the maintenance/documentation/explanatory burden here?

Reply:
I agree with Phillip that I'd be reluctant to lose the error on `__bool__`. One downside of not erroring on `__bool__` is we now have an object that compares equal to anything, which can lead to weird bugs. I once lost 2 hours because `np.dtype('f8') == None` is true (for reasons), I'd expect similarly tricky bugs to turn up if deferred objects effectively compare equal to anything.

I've pushed up an alternative solution in #8474. It may or may not solve the original ask, but I'd much prefer it to the solution here.

Reply:
Cross-posting that the underlying `Resolver` object can be retrieved which is actually hashable, see https://github.com/ibis-project/ibis/pull/8474#issuecomment-1967389714 for details.